dotnet/src/SemanticKernel.UnitTests/KernelTests.cs

```diff
@@ -345,18 +345,272 @@ public async Task ItReturnsFunctionResultsCorrectlyAsync()
 
         // Act
         var kernelResult = await kernel.RunAsync(function1, function2);
+        var functionResult1 = kernelResult.FunctionResults.First(l => l.FunctionName == "Function1" && l.PluginName == PluginName);
+        var functionResult2 = kernelResult.FunctionResults.First(l => l.FunctionName == "Function2" && l.PluginName == PluginName);
 
         // Assert
         Assert.NotNull(kernelResult);
         Assert.Equal("Result2", kernelResult.GetValue<string>());
-
-        var functionResult1 = kernelResult.FunctionResults.First(l => l.FunctionName == "Function1" && l.PluginName == PluginName);
-        var functionResult2 = kernelResult.FunctionResults.First(l => l.FunctionName == "Function2" && l.PluginName == PluginName);
-
         Assert.Equal("Result1", functionResult1.GetValue<string>());
         Assert.Equal("Result2", functionResult2.GetValue<string>());
     }
 
+    [Fact]
+    public async Task ItReturnsChangedResultsFromFunctionInvokedEventsAsync()
+    {
+        var kernel = Kernel.Builder.Build();
+
+        // Arrange
+        [SKName("Function1")]
+        static string Function1() => "Result1";
+        var function1 = SKFunction.FromNativeMethod(Method(Function1), pluginName: "MyPlugin");
+        const string ExpectedValue = "new result";
+
+        kernel.FunctionInvoked += (object? sender, FunctionInvokedEventArgs args) =>
+        {
+            args.SKContext.Variables.Update(ExpectedValue);
+        };
+
+        // Act
+        var kernelResult = await kernel.RunAsync(function1);
+
+        // Assert
+        Assert.NotNull(kernelResult);
+        Assert.Equal(ExpectedValue, kernelResult.GetValue<string>());
```

Comment:
This is invalid scenario. If `Function1` has return type `MyComplexType`, then Kernel should return `MyComplexType` independently from changes in `SKContext` in the middle of execution.

And probably we should not have the ability to change the function result using hooks, because it looks like a place for hacks and workarounds. If my function returns `MyComplexType.PropertyA = "a"`, we should not allow to change it at the end for `MyComplexType.PropertyA = "b"`. At least, I don't see the use cases for it at the moment, but maybe I'm missing something.

Reply:
This is a requirement for the current Hooks. 

A design needs to be thought on best way to internally be able to change the results thru Invoked hook.